2. Create JsonApiClient in CodaLabManager

kashizui opened the pull request:
@klopyrev 

Replaces #447 
- Introduced hardcoded translations for REST API address
- I'm going to leave the authentication procedure as it is if that's OK, since I think it will be more work than it's worth to rewrite stuff to work better with RemoteBundleClient and LocalBundleClient.

kashizui committed:
Create JsonApiClient in CodaLabManager

kashizui committed:
Hardcode REST API address translations

klopyrev commented on codalab/lib/codalab_manager.py:
```diff
@@ -454,23 +455,77 @@ def root_user_id(self):
+            addresses of the REST servers.
+            """
+            # Manually translate known bundle service address to REST address
+            from urlparse import urlparse
```
Move import up to the top.

klopyrev commented on codalab/lib/codalab_manager.py:
```diff
@@ -497,6 +552,69 @@ def client(self, address, is_cli=True):
     def cli_verbose(self):
         return self.config.get('cli', {}).get('verbose')
 
+    def get_access_token(self, address, auth_handler):
```
I did the refactoring for you: https://github.com/codalab/codalab-cli/pull/454

I really don't like unnecessary copied code. It might be some time until the cl server is completely removed and until then there's this giant block of code that's duplicated.

klopyrev commented:
LGTM, but please switch to using https://github.com/codalab/codalab-cli/pull/454

kashizui committed:
Move import

kashizui committed:
Fix JsonApiClient to use _authenticate

kashizui committed:
Fix logout and client caching

kashizui commented:
@klopyrev fixed import, rebased on changes introduced by #454, fixed logout to delete credentials for REST client, and fixed behavior of client cache, and fixed web terminal.

kashizui commented:
@klopyrev Do you mind taking another quick look? I fixed a couple additional bugs I caught.

Most significantly, moved the code for executing commands in legacy.py from BundleService to RemoteBundleService, and instantiated a JsonApiClient there as well, so the web terminal works now too with REST calls. (tested on the groups API branch).

kashizui committed:
Fix web terminal functionality

klopyrev commented on codalab/rest/legacy.py:
```diff
@@ -383,6 +326,75 @@ def upload_bundle(self, source_file, bundle_type, worksheet_uuid):
             raise
         return new_bundle_uuid
 
+    # Create an instance of a CLI.
```
Did you not read the TODO comments above? You're making this code very confusing, especially for someone who doesn't know that this function didn't used to be here before.

klopyrev commented on codalab/rest/legacy.py:
```diff
@@ -383,6 +326,75 @@ def upload_bundle(self, source_file, bundle_type, worksheet_uuid):
+            temporary=True,
+            clients={'local': self.client},
+            rest_clients={self._rest_url(): self.rest_client})
+        manager.set_current_worksheet_uuid('local', worksheet_uuid)
```
Why local? Very confusing.

klopyrev commented on codalab/lib/codalab_manager.py:
```diff
@@ -113,14 +114,15 @@ class CodaLabManager(object):
     temporary: don't use config files
     '''
-    def __init__(self, temporary=False, clients=None):
+    def __init__(self, temporary=False, clients=None, rest_clients=None):
```
This class is a giant mess and you've made it worse. See if you can make it cleaner.

klopyrev commented:
This seems very messy, especially the manager stuff. Can you make it cleaner?

kashizui commented:
OK, I'll see what I can do. This really is an epic mess.

kashizui committed:
Fix web terminal take two

 - Move create_cli and friends back into BundleService
 - REST and old address namespaces don't clash so use one client cache
 - Factor out address translation

kashizui commented:
@klopyrev I took another shot at it:
- Moved create_cli and friends back into BundleService
- REST and old address namespaces don't clash so using one client cache now
- Factored out address translation

klopyrev commented on codalab/rest/legacy.py:
```diff
@@ -307,6 +357,7 @@ def get_user_info(self, user_id):
+
 class RemoteBundleService(object):
     '''
     Adapts the RemoteBundleClient for REST calls.
```
You obviously didn't get the hint to update the documentation for this class.

klopyrev commented on codalab/rest/legacy.py:
```diff
@@ -227,11 +264,24 @@ def get_top_level_contents(self, target):
+                self._rest_url(): rest_client
+            })
+        manager.set_current_worksheet_uuid('local', worksheet_uuid)
+        manager.config = local.config
```
Pass in the config to the constructor.

klopyrev commented:
LGTM

kashizui committed:
Pass config in CodaLabManager init
